### ISSUE TITLE
Feature Request: Ability to "find" text

### ISSUE THREAD

**Issue by sepsol** [squidowl/halloy#252]
Possibly using Regex as well

**Comment by zxrfire** [comment]
I agree with this feature. Please add ctrl+f capability to halloy like other irc clients has, makes it easier to search.

**Comment by keypush3r** [comment]
I'm planning on implementing this

**Comment by gurgl** [comment]
I'm working on this feature. Please tell me to stop if anyone is already working on it.

**Comment by tarkah** [comment]
> I'm working on this feature. Please tell me to stop if anyone is already working on it.

Awesome @gurgl! Btw we have a way to scroll to messages now checkout scroll_to_message in scroll_view.

So you could use this against a list of found messages from the search text, with next / prev functionality.

I'd recommend exposing something on the History API, such as `find_text` which returns a vec of message references

**Comment by gurgl** [comment]
> I'd recommend exposing something on the History API, such as find_text which returns a vec of message references. Store that in scroll view state along with current index into that vec for next / prev and then invalidate / refind it each time they enter text

Sweet, I'll  take a look on those parts

**Comment by casperstorm** [comment]
@tarkah and i talked about this yesterday. I think the right approach here is to create another internal buffer called Find which you can lookup text across all other buffers in.

**Comment by gurgl** [comment]
I tried to get in touch with @tarkah in #halloy several times but I never got any replies. Maybe he missed them or had a too busy schedule. I implemented quite a bit of a solution and found it was hard to extend a generic solution to all types of buffers. Especially for the rendering of the highlite in all these types of context as every buffer type had it's own rendering function iirc. This was using a copy the hexchat approach to search. Atm I've got my schedule filled up to the brim so I'm afraid I won't be jumping back at this anytime soon.

**Comment by tarkah** [comment]
Hey @gurgl apologies for not being available on IRC, I'm not on it much. Thanks for taking a look into it. I have a pretty good mental model for how to accomplish this within our data model. I'll try to write up a small RFC which can be used as a reference. Maybe it's something I can tackle if I get some more time as well.

**Comment by lorenzosu** [comment]
Hi all, just chiming in to support this feature request, a CTRL+F search (in a specific buffer / channel) would be really useful. For example I connect via ZNC and am in a few channels. Some channels are ask and 'hang around' for support and questions (also due to time zone differences, etc.) so it's really useful to even just search for my own nick in the buffer history to see if someone replied when I was maybe AFK ;-)

**Comment by lc-guy** [comment]
Agreed :] I've been looking into switching to halloy from hexchat but this is the one thing still keeping me shackled. @tarkah (sorry for the ping) did you manage to get an RFC written up for this feature?

**Comment by furudean** [comment]
This should integrate with `draft/chathistory` CAP supported networks. Local log storage may not match all the history that's accessible to the user

**Comment by kasper93** [comment]
> This should integrate with `draft/chathistory` CAP supported networks. Local log storage may not match all the history that's accessible to the user

first lets have ability to at least search local history...

**Comment by furudean** [comment]
Mhm. Just wanted to note a complete solution would have both

**Comment by MrMcEpic** [comment]
I'm happy to collaborate on this if anyone is currently working on it to make it happen.  Feel free to get in contact.

**Comment by EvansJahja** [comment]
https://github.com/squidowl/halloy/pull/1880

Work in progress, basic function should be usable

**Comment by sebbu2** [comment]
HexChat has a `/lastlog` command that open a new tab with all lines from current channel/query/server/notice/snotice/whatever containing said words (in parameters).

**Comment by casperstorm** [comment]
I'll address this once we have sqlite backend.
